### ISSUE TITLE
BaseObject and Component wrongly get/set property w/ protected (g/s)etter via Magic Method (possible security issue)

### ISSUE THREAD

**Issue by belisoful** [yiisoft/yii2#19821]
### What steps will reproduce the problem?
Make a subclass of Component with a private property and protected getter and protected setter.
```
$obj->getMyProperty();    // will fail
$obj->setMyProperty("new value");   //will fail

$value = $obj->MyProperty;   //is successful and does not fail
$obj->MyProperty = "new value";   //is successful and does not fail
```

The problem is that method_exists disregards the public/protected/private status of methods and only says if they exist.  A further step must be done to determine if the method is public or not.

### What is the expected result?
To conform to OOP principles where the property is inaccessible when the getter and setter methods are protected.

### What do you get instead?
rather than failure, errors, or NOOP (calling protected methods is NOOP), it produces the MyProperty and sets MyProperty on protected methods by magic method.

### Additional info
We found this bug in PRADO, the parent of Yii.  I thought I'd be a good Human Being(tm) and report that this could be a serious bug.

Publicly accessing protected and private properties by magic methods could be a CRITICAL security issue for some people.

**Comment by belisoful** [comment]
The price of making all properties accessible by magic method to one's own class, is making them publicly available to all.  Contrary, Limiting the public to only public getters/setters limits the class itself being able to read its own protected and private properties.

unless the calling object can be found.  then when the calling object is $this, then allow the protected/private method call.

This restores OOP principles.

**Comment by belisoful** [comment]
https://github.com/pradosoft/prado/issues/898

**Comment by bizley** [comment]
Is there a noticeable impact on the performance on your side?

**Comment by belisoful** [comment]
it was about 10 ms slower on an 8.81s full run test.

**Comment by My6UoT9** [comment]
method_exists from php docu, does say it checks if the method exists. Imho it does what it says.

**Comment by belisoful** [comment]
@My6UoT9 right, and there should be checks on access for public vs protected vs private via magic method.

**Comment by My6UoT9** [comment]
@belisoful maybe it should be named differently then, method_accessible or something like that?

**Comment by belisoful** [comment]
All we had to do was change "method_exists" to "Prado::method_exists".  It was assumed that there was some visibility validation when there was none.  Our method does the visibility validation.  simple switch if they do stay the same.

**Comment by My6UoT9** [comment]
It feels wrong, cause the method does exist, but is not accessible from outside. If one already switches to a static method, it can have any name.

**Comment by belisoful** [comment]
and your comment is one reason I raised the issue here.  for that feed back.

**Comment by notaidea** [comment]
getter/setter method must be public，otherwise there is no way to access a specified field.

other programing languages(e.g c#) that support "property" originally，the compiler generates public getter/setter method and method-name with a random string or prefix，you don't use the getter/setter method conveniently,but it's really public.

**Comment by belisoful** [comment]
@notaidea If there is a public getter but not a public setter (or visa versa), using the magic method will still access the protected/private method.  That is a violation of OOP.  

the idea that setters and getters "should" or "must" be public is not valid.  typically they are, yes, but "must" is way way too strong.

At PRADO have a protected setter on at least one object with a public getter: TList::ReadOnly.  I don't think you have list classes in Yii any more, so it doesn't translate but serves as an example.  We don't want people just changing the ReadOnly status of the list, however subclasses can reset the ReadOnly on rare occasions for internal purposes.   It is unacceptable that ReadOnly setter is protected directly, but magic methods allow setter access.

**Comment by belisoful** [comment]
I suggest you run a `grep "protected function set" -r *` and see how much is incorrectly visible by magic methods.

**Comment by belisoful** [comment]
Also, because PHP refers to "public, protected, private" as Visibility, the name I'm going with is Method_Visible.

**Comment by samdark** [comment]
See https://3v4l.org/0S3iC

**Comment by DeryabinSergey** [comment]
> 3. Can it be done with `is_callable([$object, $method])`?

`is_callable` will being always return `true` for protected and self private methods, because we will calling `is_callable` from object environment.

> See https://3v4l.org/0S3iC

It\`s works, when called outside object, not with `$this`

**Comment by samdark** [comment]
Well, could be solved like this: https://3v4l.org/aCl1r See `checkOutsideContextFromInside()`.

**Comment by DeryabinSergey** [comment]
> is way faster than reflection + backtracing.

I thin so. This is one of the most frequently used features.

@samdark what do you think about next case: https://3v4l.org/eIL3A We can use protected method inside class environment

**Comment by samdark** [comment]
That's called "friendly classes" and it is fine. It works the same way as PHP works for classes.

**Comment by DeryabinSergey** [comment]
Not really. When I access a property of the parent class in my example, I expect the protected method that is available to be called. But it is not accessible from the outside and I get an error

**Comment by belisoful** [comment]
> Well, could be solved like this: https://3v4l.org/aCl1r See `checkOutsideContextFromInside()`. Liklely that is way faster than reflection + backtracing

Within the magic method, the methods are accessible by is_callable when they shouldn't be.  they shouldn't be callable from the caller via magic method.  Is PHP visibility compliance worth a few milliseconds of processing time for a commonly used feature?

**Comment by samdark** [comment]
> Is PHP visibility compliance worth a few milliseconds of processing time for a commonly used feature?

Could be since it is basically everywhere in Yii2.

**Comment by belisoful** [comment]
For being everywhere, only a few milliseconds isn't half bad....   accessing protected and private functions by magic method is a common error developers make.  Events could access a behaviors protected/private method as event handler as well.

 and if Yii2 (magic methods) can't conform to PHP visibility standards, I question Yii2.

**Comment by samdark** [comment]
I'm not saying that we should not fix it. I'm insisting on doing the benchmarks before applying any solution and checking if the performance drop is significant.

**Comment by belisoful** [comment]
I entirely agree.

**Comment by samdark** [comment]
@belisoful do you have some time to prepare a pull request with some tests and the code changes (and maybe a benchmark)? Having the tests we can optimize the code without missing edge cases.

**Comment by belisoful** [comment]
I'm busy writing a new package for Prado with unit tests and such.

This is the full scale update we have at Prado for this: https://github.com/belisoful/prado/commit/6296b7eb5185c118ab3ad8a8ed6e3b1cd791ab9c

It includes the unit tests you are looking for.  it includes code changes, just not for Yii.  This does not include the final name `method_visible`.

**Comment by belisoful** [comment]
I _can_ but it'll take a while.

**Comment by samdark** [comment]
That's fine. We don't have resources to work on it now as we're busy with Yii3 and it doesn't seem to be critical for current Yii2.

**Comment by belisoful** [comment]
Indeed, what new features does Yii3 have?

**Comment by WarLikeLaux** [comment]
Took a look and ran the numbers, since performance was the open question.

Reproduced it on plain `Component` first: a `protected getToken()/setToken()` is fully reachable through magic (`$obj->token` reads and writes it), while direct `$obj->getToken()` throws. `hasProperty()`/`canGetProperty()` return `true` too, ignoring visibility.

Gate cost on the `__get` hot path (PHP 8.5, public getter, best of 5 over 2M calls):

```
method_exists (current)              206 ns/op   1.0x
ReflectionMethod::isPublic + cache   346 ns/op   1.7x
unscoped is_callable                 386 ns/op   1.9x
bindTo() prototype (aCl1r)           586 ns/op   2.8x
```

The `bindTo` trick is actually the slowest here, below cached reflection. Any gate roughly doubles the isolated cost (less in a real `__get` with behaviors around it; and this is 8.5, 7.4 is relatively worse for reflection/closures).

Speed isn't the real fork though. The cheap checks (`isPublic`, unscoped `is_callable`) all mean the same thing: "public-only via magic, for everyone". That breaks a subclass reaching a parent's protected setter through a property, which is @DeryabinSergey's eIL3A case where `$this->two = 2` from the child throws. Keeping native PHP semantics (the child does have access) needs `debug_backtrace` to see the real caller, and it only fires for non-public methods, so the public hot path stays at the `isPublic` cost.

So before a PR would make sense:

1. Which semantics for 2.0: strict "public-only via magic" (cheap, but a BC break for friendly access on its own), or native with a backtrace fallback?
2. There are ~150 `protected get*/set*` in the framework today, some reachable via magic, so either way this is a real 2.0 BC. Acceptable?

Can build it out with tests and a proper benchmark once the semantics are picked.
